crates/flox-rust-sdk/src/models/environment/path_environment.rs

```diff
@@ -398,83 +403,66 @@ impl PathEnvironment<Original> {
     pub fn open(
-        path: impl AsRef<Path>,
+        pointer: PathPointer,
+        dot_flox_path: impl AsRef<Path>,
         temp_dir: impl AsRef<Path>,
     ) -> Result<Self, EnvironmentError2> {
-        let path = path.as_ref().to_path_buf();
-        let dot_flox_path = path.join(".flox");
-        let env_dir_entry = dot_flox_path
-            .read_dir()
-            .map_err(EnvironmentError2::ReadDotFlox)?
-            .find_map(|entry| {
-                let entry = entry.ok()?;
-                let path = entry.path();
-                if path.is_dir() {
-                    Some(entry)
-                } else {
-                    None
-                }
-            })
-            .ok_or(EnvironmentError2::EnvNotFound)?;
-
-        let env_path = env_dir_entry
-            .path()
+        let env_dir = dot_flox_path.as_ref().join(ENVIRONMENT_DIR_NAME);
+        if !env_dir.exists() {
+            Err(EnvironmentError2::EnvNotFound)?;
+        }
+
+        let env_path = env_dir
             .canonicalize()
             .map_err(EnvironmentError2::EnvCanonicalize)?;
+
+        // TODO: replace with checks for manifest file once we transition
         if !env_path.join("flake.nix").exists() {
             Err(EnvironmentError2::DirectoryNotAnEnv)?
         }
 
-        let name = EnvironmentName::from_str(&env_dir_entry.file_name().to_string_lossy())?;
-
         Ok(PathEnvironment {
             path: env_path,
-            name,
+            pointer,
             temp_dir: temp_dir.as_ref().to_path_buf(),
             state: Original,
         })
     }
 
-    /// Find the closest `.flox` starting with `current_dir`
-    /// and looking up ancestor directories until `/`
-    pub fn discover(
-        current_dir: impl AsRef<Path>,
-        temp_dir: impl AsRef<Path>,
-    ) -> Result<Option<Self>, EnvironmentError2> {
-        let maybe_dot_flox = current_dir
-            .as_ref()
-            .ancestors()
-            .find(|ancestor| ancestor.join(".flox").exists());
-
-        let with_dot_flox = if let Some(with_dot_flox) = maybe_dot_flox {
-            with_dot_flox
-        } else {
-            return Ok(None);
-        };
-
-        Some(Self::open(with_dot_flox, temp_dir)).transpose()
-    }
-
     /// Create a new env in a `.flox` directory within a specific path or open it if it exists.
     ///
     /// The method creates or opens a `.flox` directory _contained_ within `path`!
     pub fn init(
+        pointer: PathPointer,
         path: impl AsRef<Path>,
-        name: EnvironmentName,
         temp_dir: impl AsRef<Path>,
     ) -> Result<Self, EnvironmentError2> {
-        if Self::open(&path, &temp_dir).is_ok() {
-            Err(EnvironmentError2::EnvironmentExists)?;
+        match EnvironmentPointer::open(path.as_ref()) {
+            Err(EnvironmentError2::DirectoryNotAnEnv) => {},
+            Err(e) => Err(e)?,
+            Ok(_) => Err(EnvironmentError2::EnvironmentExists)?,
```

Comment:
suggestion nonblocking: I think it would be more straightforward to error unconditionally if `dot_flox_path.exists()`

Reply:
Yeah, that's ok for now. I thought of `.flox` to possibly contain more than the environment. But indeed at this time we just consider an environment present there.
The question is still where we want to handle the pointer stuff cleanly -- its not blocking though, lets bikeshed when we have time.